### ISSUE TITLE
sporadic 503 error for https://login.anosrep.org/wsapi/auth_with_assertion

### ISSUE THREAD

**Issue by jrgm** [mozilla/persona#3309]
I've seen this _once_ in my own browser, and grepping on stage shows this has happened to 12% of requests to auth_with_assertion.

STR
1. Just go through a normal flow to sign up as a new primary user (eyedee.me, yahoo.com).

When the error occurs, the log messages in browserid.log are:

```
{"level":"info","message":"successfully created primary acct for 2y8qk101ir@eyedee.me (11609398)","timestamp":"2013-04-25T23:32:39.137Z"}
{"level":"warn","message":"database is down, cannot process request: Error: authenticateSession called with undefined lastPasswordReset","timestamp":"2013-04-25T23:32:39.139Z"}
{"level":"info","message":"127.0.0.1 - - [Thu, 25 Apr 2013 23:32:39 GMT] \"POST /wsapi/auth_with_assertion HTTP/1.1\" 503 41 \"https://login.anosrep.org/signin\" \"Mozilla/5.0 (Windows NT 6.1; WOW64; rv:16.0) Gecko/20100101 Firefox/16.0\"","timestamp":"2013-04-25T23:32:39.140Z"}
```

In former scl2 stage, since the beginning for March the URL was been called thousands of times with 100% success.

(Perhaps there is greater network "distance" between the db master and the slaves, and we are racing with replication between the creation of the user row by the dbwriter /wsapi/create_account_with_assertion, and the read from a slave with wsapi.authenticateSession. Unfortunately, ping is not available, and I'm not sure how I compute replication lag from pt-heartbeat, but I'll look into that).

**Comment by karlht** [comment]
I have also replicated this (just once) in a normal manual browser login flow.

**Comment by mostlygeek** [comment]
Here is what it looks like: 

![wgfK](https://f.cloud.github.com/assets/83972/433182/19e0c0be-aec1-11e2-8d18-015eb07d46ee.png)

**Comment by jrgm** [comment]
One point of clarification. When I hit my error, I looked in the database slaves and the row was there. The error would have only occurred if the row does not exist.

**Comment by lloyd** [comment]
We need someone to dig into code and enumerate the conditions under which this can occur.  This might lead to reliable steps to reproduce.

-- lloyd (thumb-typing)

**Comment by lloyd** [comment]
We could also write some code to exercise this flow programmatically (against apis) to vet @jrgm 's timing theory - which sound incredibly plausible.

**Comment by jrgm** [comment]
I went back over logs in stage since 04/24. While load was running, there was an overall 8% failure rate. In recent days, with no load_gen running, there is maybe a 1% failure rate. Which is consistent with this being a race (without load, there is not much, if any, backlog of queries in the replication queue).

**Comment by jrgm** [comment]
Fix from train-2013.04.26 (GH-3322) and dev branch (GH-3321) have been merged. Closing.

**Comment by jrgm** [comment]
Sorry. Cleaning up some stuff that I had already looked at, but not marked done. Not seeing this error occur with train-2013.05.22 (or 05.08). Verified.
